File: backend/app/schemas/order.py
from pydantic import Field
from app.schemas.to_camel import BaseConfigModel
from app.schemas.user import UserAddressOut, UserProfileOut
from app.db.models.order import OrderStatus
from datetime import datetime
from typing import Optional, List
from decimal import Decimal


class OrderItemCreate(BaseConfigModel):
    product_id: int = Field(gt=0)
    variant_id: int = Field(gt=0)
    quantity: int = Field(gt=0)


class OrderCreate(BaseConfigModel):
    address_id: int = Field(gt=0)
    items: List[OrderItemCreate] = Field(default_factory=list)


class OrderItemOut(OrderItemCreate):
    id: int = Field(gt=0)
    name: str = Field(min_length=1, max_length=100)
    price_at_purchase: Decimal = Field(gt=0)
    image_url: Optional[str] = Field(default=None)
    subtotal: Decimal = Field(gt=0)


class OrderOut(BaseConfigModel):
    id: int = Field(gt=0)
    status: OrderStatus
    total_amount: Decimal = Field(gt=0)
    created_at: datetime
    updated_at: Optional[datetime] = Field(default=None)
    payment_status: Optional[str] = None
    payment_method: Optional[str] = None
    transaction_id: Optional[str] = None
    order_id_midtrans: Optional[str] = None
    snap_token: Optional[str] = None
    snap_redirect_url: Optional[str] = None
    address: UserAddressOut
    user: UserProfileOut
    items: List[OrderItemOut] = Field(default_factory=list)

class MidtransTransactionOut(BaseConfigModel):
    order_id: str = Field(gt=0)
    token: str = Field(min_length=1)
    redirect_url: str = Field(min_length=1)
    expired_at: datetime

File: src/api/order.ts
import axios, { AxiosError } from "axios";
import { UserAddressOut, UserProfileOut } from "./user";

const API_URL = process.env.NEXT_PUBLIC_API_URL;

interface ApiErrorResponse {
  detail?: string;
  message?: string;
}

export type OrderStatus =
  | "all"
  | "pending"
  | "confirmed"
  | "shipped"
  | "delivered"
  | "cancelled"
  | "refunded";

export interface OrderItemCreate {
  productid: number;
  variantId: number;
  quantity: number;
}

export interface OrderCreate {
  addressId: number;
  items: OrderItemCreate[];
}

export interface MidtransTransactionOut {
  orderId: string;
  token: string;
  redirectUrl: string;
  expiredAt: string;
}

export interface OrderItemOut {
  id: number;
  productId: number;
  quantity: number;
  priceAtPurchase: number;
  name: string;
  subtotal: number;
  imageUrl?: string;
}

export interface OrderOut {
  id: number;
  status: OrderStatus;
  totalAmount: number;
  createdAt: string;
  updatedAt?: string;
  midtransTransaction?: MidtransTransactionOut;
  address: UserAddressOut;
  user: UserProfileOut;
  items: OrderItemOut[];
}

export type OrderListOut = OrderOut[];

export async function createOrder(data: OrderCreate): Promise<OrderOut> {
  try {
    const res = await axios.post<OrderOut>(
      `${API_URL}/api/orders/create`,
      data,
      {
        withCredentials: true,
      }
    );
    return res.data;
  } catch (err) {
    const error = err as AxiosError<ApiErrorResponse>;
    if (error.response && error.response.data) {
      const errorData = error.response.data;
      throw new Error(errorData.detail);
    } else {
      throw new Error(error.message);
    }
  }
}

export async function orderPay(
  orderId: number
): Promise<MidtransTransactionOut> {
  try {
    const res = await axios.post<MidtransTransactionOut>(
      `${API_URL}/api/orders/${orderId}/pay`,
      null,
      {
        withCredentials: true,
      }
    );
    return res.data;
  } catch (err) {
    const error = err as AxiosError<ApiErrorResponse>;
    if (error.response && error.response.data) {
      const errorData = error.response.data;
      throw new Error(errorData.detail);
    } else {
      throw new Error(error.message);
    }
  }
}

export async function getUserOrder(): Promise<OrderListOut> {
  const res = await axios.get<OrderListOut>(`${API_URL}/api/orders/me`, {
    withCredentials: true,
  });
  return res.data;
}
